rnnalyse/extractors/base_extractor.py

```diff
@@ -147,38 +145,76 @@ def _print_time_info(self, start_time: float, print_every: int) -> None:
               f'Time: {minutes:>3.0f}m {seconds:>2.2f}s\t'
               f'Speed: {speed:.2f}s/sen')
 
-    def _extract_sentence(self, sentence: Sentence) -> None:
+    def _extract_sentence(self, sentence: LabeledSentence) -> int:
         """ Generates the embeddings of a sentence and writes to file.
 
         Parameters
         ----------
         sentence : Sentence
             The to-be-extracted sentence, represented as a list of strings.
+
+        Returns
+        -------
+        num_extracted: int
+            Number of extracted activations for this sentence.
         """
+
+        sen_activations: PartialActivationDict = self._init_sen_activations(0)
+
+        activations: FullActivationDict = self.init_lstm_states.states
+
+        for i, token in enumerate(sentence.sen):
+            out, activations = self.model(token, activations)
+
+            # Check whether current activations match criterion defined in selection_func
+            if self.selection_func(i, token, sentence):
+                for l, name in self.activation_names:
+                    sen_activations[(l, name)] = np.append(
+                        sen_activations[(l, name)], activations[l][name].detach().numpy()[np.newaxis, ...], axis=0
+                    )
+
+        num_extracted = list(sen_activations.values())[0].shape[0]
+        self._dump_activations(sen_activations)
+
+        return num_extracted
+
+    def extract_average_eos_activations(self):
```

Comment:
There is a lot of repetition here wrt to the `extract` function. If you'd move the selection function from init to the `extract` function, I think you could just reuse it to extract activations and then process them afterwards, right?

Reply:
Very good idea!